File: generator_web_app/src/modules/api/surveys.ts
// API methods for managing survey objects
// Author: Michael B. Lance

// ---------------------------------------------------------------------------------------------------------------------------

import { HerdUnit, Survey, Image } from '@/types/generatorobjects.ts';
import type { HerdUnitIntf, ImageIntf, imageRecords, SurveyIntf } from '@/types/generatorobjects.ts';
import { ApiError } from '@/modules/api/errors.ts'
import { api_url } from '@/modules/api/apiV1Methods.ts';

// ---------------------------------------------------------------------------------------------------------------------------

export async function getSurveyHerdUnits(survey_id: string): Promise<HerdUnit[]> {
  const response = await fetch(`${api_url}/surveys/${survey_id}/herd-units`, {
    method: 'GET',
    credentials: 'include',
    headers: {
      'Content-Type': 'application/json',
    },
  });
  if (!response.ok) throw new ApiError(await response.json());
  const resp = await response.json();
  let herd_units = [];
  for (const herd_unit of resp) herd_units.push(new HerdUnit(herd_unit as HerdUnitIntf));
  return herd_units;
}

// ---------------------------------------------------------------------------------------------------------------------------

export async function getSurveyImages(survey_id: string, page: number, per_page: number): Promise<imageRecords> {

  const response = await fetch(`${api_url}/surveys/${survey_id}/images?page=${page}&per_page=${per_page}`, {
    method: "GET",
    credentials: "include",
    headers: {
      "Content-Type": "appication/json",
    },
  });

  if (!response.ok) throw new ApiError(await response.json());

  const resp = await response.json();

  let image_records = {
    total_images: resp.total,
    images: {}
  } as imageRecords;

  for (const image of resp.images) image_records.images[image.uuid] = new Image(image as ImageIntf);
  return image_records;
}

// ---------------------------------------------------------------------------------------------------------------------------

export interface createSurveyOptions {
  "name": string;
  "project_id": string;
  "herd_unit_id": string;
  "survey_date": string;
  "additional_info": string;
}

export async function createSurvey(options: createSurveyOptions): Promise<Survey> {
  const response = await fetch(`${api_url}/surveys`, {
    method: 'POST',
    credentials: 'include',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(options),
  });
  if (!response.ok) throw new ApiError(await response.json());
  const resp = await response.json();
  let survey = new Survey(resp as SurveyIntf);
  return survey;
}

// ---------------------------------------------------------------------------------------------------------------------------

export async function deleteSurvey(survey_id: string): Promise<boolean> {
  const response = await fetch(`${api_url}/surveys/${survey_id}`, {
    method: "DELETE",
    credentials: "include",
    headers: {
      "Content-Type": "application/json"
    },
  });

  if (!response.ok) throw new ApiError(await response.json());

  return true;
}

File: airial_api/app/routers/surveys.py
# Endpoints for managing images in the API
# Author: Michael B. Lance

# ---------------------------------------------------------------------------------------------------------------------------

from flask import Blueprint, abort, request, current_app
from psycopg import DatabaseError
from app.decorators import permission_required
from database.errors import AuthenticationFailure, AuthorizationFailure, ObjectNotFound
from database.object_models.project_management import CreateSurveyReq, UpdateSurveyReq
from app.extensions import base
from datetime import datetime
from flask_pydantic import validate
from flask_login import (
    login_required,
)
from uuid import UUID

from database.object_models.project_management.surveys import SurveyImageQuery

surveyBp = Blueprint("surveys", __name__, url_prefix="/api/v1/surveys")

# ---------------------------------------------------------------------------------------------------------------------------
# GET


@surveyBp.get("/<string:survey_id>")
@login_required
@permission_required("access")
def get_by_id(survey_id: str):
    """ """
    try:
        survey = base.get_survey(UUID(survey_id))
    except ObjectNotFound as e:
        current_app.logger.exception(e)
        abort(404, str(e))
    except AuthenticationFailure as e:
        current_app.logger.exception(e)
        abort(401, str(e))
    except (DatabaseError, Exception) as e:
        current_app.logger.exception(e)
        abort(500)

    return survey.to_dict(), 200


# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


@surveyBp.get("/<string:survey_id>/annotations")
@login_required
@permission_required("access")
def get_annotations(survey_id: str):
    """ """
    try:
        annotations = base.get_survey_annotations(UUID(survey_id))
    except ObjectNotFound as e:
        current_app.logger.exception(e)
        abort(404, str(e))
    except AuthorizationFailure as e:
        current_app.logger.exception(e)
        abort(401, str(e))
    except (DatabaseError, Exception) as e:
        current_app.logger.exception(e)
        abort(500)

    return [annotation.to_dict() for annotation in annotations], 200


# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


@surveyBp.get("/<string:survey_id>/herd-units")
@login_required
@permission_required("access")
def get_herd_units(survey_id: str):
    """ """
    try:
        herd_units = base.get_survey_herd_units(UUID(survey_id))
    except ObjectNotFound as e:
        current_app.logger.exception(e)
        abort(404, str(e))
    except AuthorizationFailure as e:
        current_app.logger.exception(e)
        abort(401, str(e))
    except (DatabaseError, Exception) as e:
        current_app.logger.exception(e)
        abort(500)

    return [herd_unit.to_dict() for herd_unit in herd_units], 200


# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


@surveyBp.get("/<string:survey_id>/images")
@login_required
@permission_required("access")
@validate()
def get_images(query: SurveyImageQuery, survey_id: str):
    """ """
    offset = (query.page - 1) * query.per_page
    try:
        images, total_images = base.get_survey_images(
            UUID(survey_id), query.per_page, offset
        )
    except (DatabaseError, Exception) as e:
        current_app.logger.exception(e)
        abort(500)

    return {
        "images": [image.to_dict() for image in images],
        "total": total_images,
        "page": query.page,
        "per_page": query.per_page,
    }, 200


# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


# TODO: Update to use pydantic query class
@surveyBp.get("/<string:survey_id>/annotated_images")
@login_required
@permission_required("access")
def get_annotated_images(survey_id: str):
    """ """
    date_range = request.args.get("date_range", None)
    surveys = [base.get_survey(UUID(survey_id)).survey_id]
    labels = request.args.getlist("label", type=int)
    herd_units = request.args.getlist("herd_unit", type=int)
    format_pattern = "%m/%d/%Y %H:%M:%S"
    if date_range is not None:
        date_range = (
            datetime.strptime(date_range[0], format_pattern),
            datetime.strptime(date_range[1], format_pattern),
        )

    try:
        data = base.get_survey_annotated_images(labels, date_range, surveys, herd_units)
    except ObjectNotFound as e:
        current_app.logger.exception(e)
        abort(404, str(e))
    except AuthorizationFailure as e:
        current_app.logger.exception(e)
        abort(401, str(e))
    except (DatabaseError, Exception) as e:
        current_app.logger.exception(e)
        abort(500)

    return data, 200


# ---------------------------------------------------------------------------------------------------------------------------
# POST


@surveyBp.post("")
@login_required
@permission_required("access")
@validate()
def create(body: CreateSurveyReq):
    """ """
    try:
        survey = base.create_survey(body)
    except AuthorizationFailure as e:
        current_app.logger.exception(e)
        abort(401, str(e))
    except (DatabaseError, Exception) as e:
        current_app.logger.exception(e)
        abort(500)

    return survey.to_dict(), 201


# ---------------------------------------------------------------------------------------------------------------------------
# PUT

# ---------------------------------------------------------------------------------------------------------------------------
# PATCH


@surveyBp.patch("/<string:survey_id>")
@login_required
@permission_required("access")
@validate()
def update(body: UpdateSurveyReq, survey_id: str):
    """ """
    try:
        survey = base.update_survey(UUID(survey_id), body.model_dump())
    except ObjectNotFound as e:
        current_app.logger.exception(e)
        abort(404, str(e))
    except AuthorizationFailure as e:
        current_app.logger.exception(e)
        abort(401, str(e))
    except (DatabaseError, Exception) as e:
        current_app.logger.exception(e)
        abort(500)

    return survey.to_dict(), 200


# ---------------------------------------------------------------------------------------------------------------------------
# Delete


@surveyBp.delete("/<string:survey_id>")
@login_required
@permission_required("access")
def delete_survey(survey_id: str):
    """ """

    try:
        base.delete_survey(UUID(survey_id))
    except ObjectNotFound as e:
        current_app.logger.exception(e)
        abort(404, str(e))
    except AuthorizationFailure as e:
        current_app.logger.exception(e)
        abort(401, str(e))
    except (DatabaseError, Exception) as e:
        current_app.logger.exception(e)
        abort(500)

    return "", 204
